airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io/airbyte/cdk/read/cdc/CdcPartitionsCreator.kt

```diff
@@ -41,54 +41,79 @@ class CdcPartitionsCreator<T : Comparable<T>>(
-            when (val incumbentOpaqueStateValue = feedBootstrap.currentState) {
-                null -> syntheticInput
-                else -> {
-                    // validate if existing state is still valid on DB.
-                    try {
-                        creatorOps.deserialize(incumbentOpaqueStateValue, activeStreams)
-                    } catch (ex: ConfigErrorException) {
-                        log.error(ex) { "Existing state is invalid." }
-                        throw ex
-                    } catch (_: OffsetInvalidNeedsResyncIllegalStateException) {
-                        // If deserialization concludes we need a re-sync we rollback stream states
-                        // and put the creator in a Need Restart mode.
-                        // The next round will throw a transient error to kickoff the resync
-                        feedBootstrap.stateQuerier.resetFeedStates()
-                        CDCNeedsRestart = true
-                        syntheticInput
-                    }
+        val warmStartState: DebeziumWarmStartState? =
+            feedBootstrap.currentState?.let {
+                try {
+                    creatorOps.deserializeState(it)
+                } catch (e: Exception) {
+                    // This catch should be redundant for well-behaved implementations
+                    // but is included anyway for safety.
+                    AbortDebeziumWarmStartState(e.toString())
                 }
             }
-
+        val debeziumProperties: Map<String, String>
+        val startingOffset: DebeziumOffset
+        val startingSchemaHistory: DebeziumSchemaHistory?
+        when (warmStartState) {
+            null -> {
+                debeziumProperties = creatorOps.generateColdStartProperties()
+                startingOffset = syntheticOffset
+                startingSchemaHistory = null
+            }
+            is ValidDebeziumWarmStartState -> {
+                debeziumProperties = creatorOps.generateWarmStartProperties(activeStreams)
+                startingOffset = warmStartState.offset
+                startingSchemaHistory = warmStartState.schemaHistory
+            }
+            is AbortDebeziumWarmStartState -> {
+                val e =
+                    ConfigErrorException(
+                        "Incumbent state is invalid, reason: ${warmStartState.reason}"
+                    )
+                log.error(e) { "Aborting. ${e.message}." }
+                throw e
+            }
+            is ResetDebeziumWarmStartState -> {
+                reset.set(true)
+                log.warn {
+                    "Triggering reset. " +
+                        "Incumbent state is invalid, reason: ${warmStartState.reason}."
+                }
+                debeziumProperties = creatorOps.generateColdStartProperties()
+                startingOffset = syntheticOffset
+                startingSchemaHistory = null
+            }
+        }
```

Comment:
The diff here is noisy, as it should be, but the changes really aren't that substantial.